loopy/codegen/__init__.py

```diff
@@ -550,4 +550,23 @@ def generate_body(kernel):
 
 # }}}
 
-# vim: foldmethod=marker
+from loopy.target.c import CASTIdentityMapper
+class FunctionDeclExtractor(CASTIdentityMapper):
+    def __init__(self):
+        self.decls = []
+
+    def map_function_declaration(self, node):
+        self.decls.append(node)
+        return super(self.__class__, self).map_function_declaration(node)
+
+def generate_header(kernel):
+    codegen_result = generate_code_v2(kernel)
+
+    if len(codegen_result.device_programs) != 1:
+        raise LoopyError("generate_header cannot be used on programs "
+                "that yield more than one device program")
```

Comment:
I'll admit that I mostly followed the example of generate_body.
I am coming to understand more of the internals of loopy, but a lot of it is still in the 'black box' stage!

Fixed in 52c975a